Auto Remove Diff Plugin

sean-m-sullivan opened the pull request:
### What does this PR do?
Adds a lookup plugin to aid in the process of auto remove. 
lookup plugin finds diff between two lists one from the API and another user provided. 

### How should this be tested?
Automated tests have been updated to account for python.

### Is there a relevant Issue open for this?
#239

sean-m-sullivan committed:
update references to ansible.tower to controller, update playbook to be user friendly

sean-m-sullivan committed:
Merge branch 'redhat-cop:devel' into devel

sean-m-sullivan committed:
Merge branch 'redhat-cop:devel' into devel

sean-m-sullivan committed:
remove file

sean-m-sullivan committed:
Merge branch 'redhat-cop:devel' into devel

sean-m-sullivan committed:
Merge branch 'redhat-cop:devel' into devel

sean-m-sullivan committed:
Merge branch 'redhat-cop:devel' into devel

sean-m-sullivan committed:
add diff plugin

djdanielsson commented:
looks good to me, haven't had time to test it yet.

Tompage1994 commented:
Not sure if I agree with how this plugin is structured. Should we not just take in a list of the objects and the endpoint and return the diff? Seems annoying to be required to first do your own lookup when we could easily do that as part of this?

I think I'd like to see something like this quickly thought out example be possible.

e.g.
```
- set_fact:
    org_diff: "{{ lookup('redhat_cop.controller_configuration.controller_object_diff', 'organizations', controller_organizations }}"

- include_role:
    name: redhat_cop.controller_configuration.organizations:
  vars:
    controller_organizations: "{{ org_diff }}"
    controller_state: absent

- include_role:
    name: redhat_cop.controller_configuration.organizations:
```

I'd also like to see some documentation. (possibly as an example in the plugin) of a more end to end solution like above.

sean-m-sullivan commented:
So at least my idea was that this would be combined with @djdanielsson 's PR
https://github.com/redhat-cop/tower_configuration/pull/241
But after doing some testing on Merging with awx.awx, I think I have a way to incorporate it easily in with others. as all you need to do is invoke 
```
- include_role:
    name: redhat_cop.controller_configuration.organizations:
  vars:
    controller_organizations: "{{ org_diff }}"
    controller_state: absent
```
Like you said.

djdanielsson commented:
another idea would be to insert the org_diff objects into the already created controller_organizations list with the state absent so the org role only needs to run once

Tompage1994 commented:
Yeah I was thinking that but couldn't think of a good syntax to add `state: absent` to each of the objects. (Perhaps a map but I dont know if there is anything that will do that?) and then append the two lists together

sean-m-sullivan commented:
@Tompage1994 Easiest way is an option IN this plugin, I'll tinker with that :)

djdanielsson commented:
yes it should be pretty easy to do in the plugin itself and have it return the full list of objects both to add/update and remove.

Tompage1994 commented:
Maybe also an option to include the initial list too so you can literally use the following:

```
roles:
  - name: organizations
    vars:
      controller_organizations: "{{ lookup('redhat_cop.controller_configuration.controller_object_diff', 'organizations', controller_organizations, set_absent=true, with_present=true }}"
```

I think something like that would be pretty cool

djdanielsson commented:
> ```
> roles:
>   - name: organizations
>     vars:
>       controller_organizations: "{{ lookup('redhat_cop.controller_configuration.controller_object_diff', 'organizations', controller_organizations, set_absent=true, with_present=true }}"
> ```
What would it do if both options were false? because it seems like it wouldn't do anything, so I think we should just ask with_present true or false and if false it would just return the list of items to remove.

Tompage1994 commented:
> > ```
> > roles:
> >   - name: organizations
> >     vars:
> >       controller_organizations: "{{ lookup('redhat_cop.controller_configuration.controller_object_diff', 'organizations', controller_organizations, set_absent=true, with_present=true }}"
> > ```
> 
> What would it do if both options were false? because it seems like it wouldn't do anything, so I think we should just ask with_present true or false and if false it would just return the list of items to remove.

My thoughts are that `set_absent` would add `state: absent` to the data structure returned and `with_present` would also return the list passed into the plugin. Though not sure it would make sense if `set_absent` were false and `with_present` were true.

sean-m-sullivan commented:
My thoughts are that 'set_absent' would add state: absent to the data structure returned and 'with_present' would only use those already set to present to compare. 

I think we agreed on the 'set_absent', but the 'with_present' what are we looking for exactly??? 
1. That we return all items with the absent set only on those diffed? I'm on the fence about this one, but could see where we merge the two lists somehow to get add absent to the ones that we Identify. 
2. Only use those with state absent or no state set?
3. Returning one of the lists? they are already set as vars, not sure what the return would do.

djdanielsson commented:
it would be 1. the idea is if you could return everything that you want to add and remove into one list then you could just call the role to do that once which would be more efficient. I think it should be pretty easy for the plugin to do once you have the diff list and added the state absent to each one of those objects then you could just combine that list with the given list that it compared with to have a full list of everything you want to add and remove.

sean-m-sullivan commented:
@djdanielsson If you could review this, I think I've updated it to work like you said, and agree that it makes more sense once I returned to it.

I've updated the lookup plugin, and if run in default, and prior to invoking roles, it should remove any unwanted objects. 
I've added tests for all the objects I think that should be used with, if we want more, we can look at that as well. Right now the had code is to look at the summary fields in the api for the org name. 

The testing went through a few iterations, but I like the solution presented. Its looping through a file, and using a file to describe each item. Changes were made to at least have two items of every object in the default org.

Tompage1994 commented on examples/tasks/differential.yml:
```diff
@@ -0,0 +1,14 @@
+  assert:
+    that:
+      - set_absent_diff ==  differential_item.expected_test_result
+...
```
Could you add a test case for where you have set_absent and with_present set as true and compare to 

```
differential_item.expected_test_result + differential_item.differential_test_items
```

sean-m-sullivan replied:
The 2nd test item Projects is specifically set to have both set_absent and with_present to be true, and the expected test has the relevant items already added. 
Normal default tests are with_present as false, set_absent to be True.

Tompage1994 commented on plugins/lookup/controller_object_diff.py:
```diff
@@ -0,0 +1,112 @@
+RETURN = """
+_raw:
+  description:
+    - Name and Organzization of objects that are NOT in the list
```
I suppose this is not necessarily true if you set `with_present` so you may want to add a caveat for that

sean-m-sullivan replied:
Getting both the return and examples updated, as things have changed quite a bit since I wrote that.

Tompage1994 commented on plugins/lookup/controller_object_diff.py:
```diff
@@ -0,0 +1,112 @@
+    default: True
+"""
+
+EXAMPLES = """
```
Could you add an example with the set_absent and with_present set. Maybe also add using it with another role to show a fuller usage

sean-m-sullivan replied:
@Tompage1994 A better more comprehensive example with both options has been set.

Tompage1994 reviewed (changes requested):
A couple of small things but otherwise looking good

djdanielsson commented:
looks pretty good to me

sean-m-sullivan commented on examples/configs/differential_items.yml:
```diff
@@ -0,0 +1,137 @@
+      - name: controller_application-app2
+        organization: Default
+        state: absent
+  - name: projects
```
@Tompage1994 This is the example with set_absent and with_present to be true

sean-m-sullivan commented:
@Tompage1994 should be ready for review

Tompage1994 reviewed (approved):
Looks good!
